NOTISSUES: Add kinesis tests, refactor doc and refactor code style

vir-mir opened the pull request:
1. change in tests:
    - add kinesis tests
    - remove `test_patches`, dependency management through `pyup-bot`, this test prevents him 
    - Improved `make` file, for testing locally or via AWS
2. change in docs:
   - add kinesis summary and useful links
   - common format for message `commit`
   - common format for name `branch`
   - common format for file `CHANGES.txt` 
   - added more input information in instructions for contributors
3. refactor code style:
    - add `isort` 
4. change in examples:
   - add kinesis example
   - common format for a directory with examples
5. change in requirements:
   - grouping dependencies in one directory for `pyup-bot`

@jettify review please

thehesiod commented:
You cannot remove test patches, do you understand its purpose? If it fails it means a human needs to validate the upstream changes. There may be new feature support needed.

vir-mir commented:
@thehesiod, Yes, I understand.
But this test does nothing useful, it just checks that the hash has not changed.
For this there are tests for the used functionality.
All versions of dependencies are freeze in `requirements / *`. To manage dependencies, use `pyup-bot`, in any case human intervention will be needed.

But now this test only hinders, for example https://github.com/aio-libs/aiobotocore/pull/653 the result is that all the functionality is working, according to the tests. But due to the fact that the hash has not converged, we get fail https://travis-ci.com/aio-libs/aiobotocore/jobs/164023523

I do not understand why this is, if the functionality works. Even if the user receives an error, after the update, due to the fact that we have not covered all the tests, can use roll back the version and create issues.

thehesiod commented:
that is not how releases work, we must try to guarantee that with each release we've done everything possible to ensure that the release works and behaves like the botocore equivalent. If aiobotocore were to accept turning off the patches I would start working from a fork. The patches ensure that if any changes are made to the botocore super classes a human will verify if any changes are necessary in aiobotocore. Without it we'll start to have failures creep in over time. Long term we have other strategies to remove this requirement. Please see the other open issues

vir-mir commented:
Who should make the decision on updating super classes?

thehesiod commented:
The super classes are in botocore, we need to mirror their changes in aiobotocore. I think we need to expand the readme to explain this better. I'll try to write something up

vir-mir commented:
@thehesiod but it is not. there is still a main class aiobotocore and aiohttp

thehesiod commented:
@vir-mir Like I said, this is not a standard library, you cannot write unittests for functionality you do not know exists.  A new botocore can add new functionality for which we have not added tests, and the patches test will catch this.  Sorry I haven't had a chance to write this doc yet. Hopefully soon.

vir-mir commented:
@thehesiod  you look the changes, I returned the test

vir-mir commented:
https://github.com/aio-libs/aiobotocore/pull/654/files/327b38ee3b39ec64aa4c958393197bdce6c6d35d..d227ee5fd48768fc4521d72bb8c91db83c05f370

thehesiod commented:
sorry for delay, finally updated the doc: https://github.com/aio-libs/aiobotocore/blob/master/CONTRIBUTING.rst  @vir-mir let me know what you think

thehesiod commented:
@vir-mir cool, I'll try to review this, I hope the new docs make sense, if you still think its wrong lets discuss here before this gets approved.

thehesiod commented on CONTRIBUTING.rst:
```diff
@@ -1,37 +1,214 @@
+
+.. code-block:: shell
+
+   $ cd aiobotocore
```
I think a better strategy going forward is using tox, so you can test against multiple python versions easily.  Thoughts?

vir-mir replied:
Start to "tox" is good for "CI", for development it is desirable to check on the new version. Here I agree with Andrey in his report https://www.youtube.com/watch?v=7KgihdKTWY4

thehesiod replied:
sorry, I don't understand what you're trying to say?  I don't understand russian either so that video doesn't help.  Is that the aiohttp author by chance, asvetlov?

thehesiod commented on CONTRIBUTING.rst:
```diff
@@ -1,37 +1,214 @@
+
+.. code-block:: shell
+
+   $ make test
```
again, would be nice to switch to tox

vir-mir replied:
for the `make test` you can hide just launch tests. it's just a level of abstraction

thehesiod commented on Makefile:
```diff
@@ -1,46 +1,83 @@
-	rm -rf build
-	rm -rf cover
-	rm -rf dist
+	@rm -rf `find . -name __pycache__`
```
can just do all this via a git clean no?

thehesiod replied:
try `git clean -fx`, it should remove the need for this big list

thehesiod commented on aiobotocore/endpoint.py:
```diff
@@ -223,33 +230,38 @@ def __init__(self, host,
 
     async def _send_request(self, request_dict, operation_model):
         attempts = 1
+        meta_key = 'ResponseMetadata'
```
the formatting here is explicit to match the version of this method in botocore, please ensure we match their method as much as possible to simplify changing for changes in new releases

thehesiod replied:
sounds good!

thehesiod commented on requirements/_main.txt:
```diff
@@ -0,0 +1,3 @@
+aiohttp==3.3.2
```
could you explain why this and awscli + boto3 were split out into separate files?

vir-mir replied:
My motivation is that all that is put through the `extras_require` in the file `setup.py` should be divided. The that is only one way to update or add dependencies.

`There should be one-- and it should be only oneobvious way to do it.`

thehesiod commented on CONTRIBUTING.rst:
```diff
@@ -1,37 +1,214 @@
+  8. Make pull request from github page for your branch against master branch
 
-Fist of all just clone repository::
+  9. Optionally make backport pull Request(s) for landing a bug fix
```
we don't have backports yet ;)

vir-mir replied:
I now think that this is superfluous, since I cannot exactly remember what exactly I wanted to say here. sorry i will remove.

thehesiod commented on CONTRIBUTING.rst:
```diff
@@ -1,37 +1,214 @@
+
+Go to the link and make sure your doc changes looks good.
+
+Commit message format
```
@jettify do we want to get this pedantic of naming your branches and formatting your commit messages in particular ways??  I think this needs to be discussed and not included in a PR like this that does a lot of things.

vir-mir replied:
Good point. I propose then to make a separate discussion. just well understand what is happening with the project, and the commit name should help in this

vir-mir commented on CONTRIBUTING.rst:
```diff
@@ -1,37 +1,214 @@
+
+  3. Make sure all tests passed
+
+  4. Make sure the coverage doesn't get worse.
```
I wanted to say here. You can check the coverage before use pull request

thehesiod replied:
there is no reference for what the current coverage levels are, I think this is only checkable via the CI no?

vir-mir commented:
@thehesiod  @jettify Thanks for the criticism, I answered all the questions. I will make another change.

Since I actively use this library in my work, I would like to help in its development, I hope we will work together.

vir-mir commented:
> sorry for delay, finally updated the doc: https://github.com/aio-libs/aiobotocore/blob/master/CONTRIBUTING.rst @vir-mir let me know what you think

@thehesiod  Thx!
Now it became clear to me why this test is needed and what to do to remove it!

I would only add motivation before the description. in order to immerse in context before reading.
example 
# Hashes of Botocore Code (important)
**We do not support all Botocore unit tests and so we decided to do the following.**

Because of the way aiobotocore is implemented (see Background section), it is very tightly coupled with botocore
.......

thehesiod commented:
re last cmt, feel free to update contributing in this PR

vir-mir commented:
@thehesiod  Hi.

I made changes according to the discussion.
For the convenience of code review, I broke into commits.


* return back the format code for botocore (https://github.com/aio-libs/aiobotocore/pull/654/commits/f975b6dd2e9b2fda9916f988aabe335956498f11)
* removed doubtful instructions (https://github.com/aio-libs/aiobotocore/pull/654/commits/d66cfe550856b056d047c724b4019145d0ed81f7)
* removed instructions of naming format commit and branch (https://github.com/aio-libs/aiobotocore/pull/654/commits/4fa418fb358d7544cecb3b36244bf2e862102c2b)
* edit instructions for changes in file CHANGES.txt (https://github.com/aio-libs/aiobotocore/pull/654/commits/551900c4de75c0a452907deff2aa06cd309c9099)
* sync master (https://github.com/aio-libs/aiobotocore/pull/654/commits/fef885d24961b7fd621d5ccda790da822b4b7818)
* edit notice of important specialty features in the library (https://github.com/aio-libs/aiobotocore/pull/654/commits/bda83f1eb3027fa4c760e6f7df8d72ff3fd53d58)

thehesiod commented on CONTRIBUTING.rst:
```diff
@@ -1,37 +1,163 @@
+
+  2. Make a change
+
+  3. Make sure all tests passed
```
```suggestion
  3. Make sure all tests pass
```

thehesiod commented on CONTRIBUTING.rst:
```diff
@@ -1,37 +1,163 @@
+      * Name commit
+
+   If you set the date and version, it will be the last and will be released
+   Version must be raised in your last committee
```
FYI, I've created templates for logging new issues and PRs.  Could you also in here update the template so the two match each other?

thehesiod commented on aiobotocore/client.py:
```diff
@@ -151,8 +149,8 @@ def paginate(self, **kwargs):
 
-            operation_model = self._service_model.\
-                operation_model(actual_operation_name)
+            operation_model = self._service_model.operation_model(
```
does this change match the botocore impl?

thehesiod commented on aiobotocore/endpoint.py:
```diff
@@ -324,18 +330,21 @@ def __init__(self, host,
-            response_dict, operation_model.output_shape)
+            response_dict,
+            operation_model.output_shape
+        )
```
do the above changes match the botocore impl

thehesiod commented on aiobotocore/paginate.py:
```diff
@@ -70,15 +69,17 @@ def _init_pager(self):
-                    self._previous_next_token == self._next_token:
+
+            is_not_none = self._previous_next_token is not None
+            if is_not_none and self._previous_next_token == self._next_token:
```
does these changes match botocore formatting?

thehesiod commented on aiobotocore/waiter.py:
```diff
@@ -1,11 +1,13 @@
+from botocore.waiter import Waiter, WaiterError, logger, xform_name
+
+# WaiterModel is required for client.py import
+WaiterModel = import_module('botocore.waiter').WaiterModel
```
why?

thehesiod commented on examples/kinesis/consumer.py:
```diff
@@ -0,0 +1,123 @@
+    async with client('kinesis', loop, **kwargs) as kinesis:
+        shards = await get_shards(kinesis, stream_name, save_shards)
+        while True:
+            await asyncio.sleep(0.2)
```
is there not a waiter you can use instead of sleep?

thehesiod commented on requirements/_main.txt:
```diff
@@ -0,0 +1,3 @@
+aiohttp==3.3.2
```
this should be open ended, it was before: `aiohttp>=3.3.1`

thehesiod commented on tests/test_basic_s3.py:
```diff
@@ -128,7 +129,7 @@
 @pytest.mark.xfail(raises=NotImplementedError)
 @pytest.mark.asyncio
-async def test_result_key_iters(s3_client, bucket_name,):
+async def test_result_key_iters(s3_client, bucket_name, ):
```
i think the extra comma at the end was a typo, mind removing?

terrycain commented:
This hasn't been touched for over a year, is there anything of value we want to incorperate from this pr?

thehesiod commented on tests/test_sns.py:
```diff
@@ -1,7 +1,7 @@
-import pytest
-import botocore
 
+import botocore
```
I think I understand what you're doing here, separating built-in modules from third party modules, I've been adding a comment before each section to make this clear in my code.  Could you please do the same?

thehesiod commented on tests/mock_server.py:
```diff
@@ -199,11 +200,24 @@ def sns_server():
 def sqs_server():
     host = "localhost"
-    port = 5004
+    port = get_free_tcp_port()
```
if you merge from master I've fixed this

thehesiod commented:
@terrycain ya there are new tests, normalizing of include ordering, more docs, etc.  But there are still a lot of questionable changes, outdated changes, and other misc problems.  Either @vir-mir should split out these changes into separate PRs or we start picking out ideas from this PR and giving credit.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/aio-libs/aiobotocore?pullRequest=654) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/aio-libs/aiobotocore?pullRequest=654) before we can accept your contribution.<br/><hr/>**Alexey Firsov** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/aio-libs/aiobotocore?pullRequest=654) it.</sub>

thehesiod commented:
a lot has changed :) can you please updated, thanks!
